src/FakeItEasy.Analyzer/ArgumentConstraintTypeMismatchAnalyzer.cs

```diff
@@ -0,0 +1,137 @@
+namespace FakeItEasy.Analyzer
+{
+    using System.Collections.Generic;
+    using System.Collections.Immutable;
+    using Microsoft.CodeAnalysis;
+#if CSHARP
+    using Microsoft.CodeAnalysis.CSharp.Syntax;
+#endif
+    using Microsoft.CodeAnalysis.Diagnostics;
+#if VISUAL_BASIC
+    using Microsoft.CodeAnalysis.VisualBasic.Syntax;
+#endif
+
+#if CSHARP
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+#elif VISUAL_BASIC
+    [DiagnosticAnalyzer(LanguageNames.VisualBasic)]
+#endif
+    public class ArgumentConstraintTypeMismatchAnalyzer : ArgumentConstraintAnalyzerBase
+    {
+        private static readonly ImmutableHashSet<string> SupportedArgumentConstraintProperties =
+            ImmutableHashSet.Create(
+                "FakeItEasy.A`1._",
+                "FakeItEasy.A`1.Ignored",
+                "FakeItEasy.A`1.That");
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } =
+            ImmutableArray.Create(DiagnosticDefinitions.ArgumentConstraintTypeMismatch);
+
+        protected override bool IsSupportedArgumentConstraintProperty(string fullName) =>
+            SupportedArgumentConstraintProperties.Contains(fullName);
+
+        protected override void AnalyzeArgumentConstraintCore(SyntaxNodeAnalysisContext context, SyntaxNode completeConstraint)
+        {
+            var constraintType = context.SemanticModel.GetTypeInfo(completeConstraint).Type as INamedTypeSymbol;
+            if (constraintType == null || !constraintType.IsValueType)

```

Comment:
Why should it be limited to value types only? I can think of cases where the problem would occur for reference types as well. For instance, a constraint of type `Bar` is passed to a method that accepts a `Foo`, and `Bar` isn't a subclass of `Foo`. Of course, it would be a compilation error anyway... unless there's an implicit conversion. But we would never match the constraint, because the method would actually receive a `Foo`, not a `Bar`.

Reply:
I'd been excluding it because of the compilation error. Do you think it's worthwhile duplicating the errors when the compiler will already be sad?

I'd not thought of an implicit conversion. This is an excellent point. I will pursue.

In the meantime, I've merged the analyzers (and tests, because once we had one analyzer with 2 diagnostics, many of the tests fought), but am lagging on the VB tests. Updates to trickle in over the next few days.

Reply:
> Do you think it's worthwhile duplicating the errors when the compiler will already be sad?

Probably not

> In the meantime, I've merged the analyzers (and tests, because once we had one analyzer with 2 diagnostics, many of the tests fought), but am lagging on the VB tests. Updates to trickle in over the next few days.

👍